File: EasyMeasure/MeasuringViewController.m
//
//  MeasuringViewController.m
//  Measure
//
//  Created by Jack on 7/12/13.
//  Copyright (c) 2013 CosmicByte. All rights reserved.
//

#import "MeasuringViewController.h"

@interface MeasuringViewController ()

@end

@implementation MeasuringViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    self.firstUpdate = YES;
    self.metric = [[NSUserDefaults standardUserDefaults] boolForKey:@"metric"];
    self.straight = [[NSUserDefaults standardUserDefaults] boolForKey:@"straight"];

    self.locationManager = [[CLLocationManager alloc] init];
    [self.locationManager setDelegate:self];
    [self.locationManager setDistanceFilter:kCLDistanceFilterNone];
    [self.locationManager setDesiredAccuracy:kCLLocationAccuracyBest];
    [self.locationManager startUpdatingLocation];
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations {
    CLLocation *location = [locations lastObject];

    if (-[location.timestamp timeIntervalSinceNow] > 2.0) {
        return;
    }

    if (location != nil) {
        if (self.straight) {
            if (self.firstUpdate) {
                self.firstUpdate = NO;
                self.firstLocation = location;
                self.distance = 0.0;
            } else {
                self.secondLocation = location;
                self.distance = [self.firstLocation distanceFromLocation:self.secondLocation];
            }
        } else {
            if (self.firstUpdate) {
                self.firstUpdate = NO;
                self.firstLocation = location;
                self.distance = 0.0;
            } else {
                self.secondLocation = location;
                self.distance += [self.firstLocation distanceFromLocation:self.secondLocation];
                self.firstLocation = location;
            }
        }
    }
    self.estimateText.text = [NSString stringWithFormat:@"%.2f %s", self.distance, (self.metric ? "m." : "yds.")];
    self.oldDistance = self.distance;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    ResultsViewController *destination = segue.destinationViewController;
    [destination sendDistance:self.distance];
}

@end

File: Easy Measure/views/MeasuringViewController.swift
//
//  MeasuringViewController.swift
//  Easy Measure
//
//  Created by Jack Cook on 6/4/14.
//  Copyright (c) 2014 CosmicByte. All rights reserved.
//

import UIKit
import CoreLocation

class MeasuringViewController: UIViewController, CLLocationManagerDelegate {

    @IBOutlet var estimateText: UILabel!
    @IBOutlet var stopButton: UIButton!
    
    var locationManager = CLLocationManager()
    var firstLocation = CLLocation()
    var secondLocation = CLLocation()
    var oldDistance = 0.0
    var distance = 0.0
    var firstUpdate = true
    var metric = NSUserDefaults().boolForKey("metric")
    var straight = NSUserDefaults().boolForKey("straight")
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        locationManager.delegate = self
        locationManager.distanceFilter = kCLDistanceFilterNone
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.startUpdatingLocation()
    }
    
    func locationManager(manager:CLLocationManager!, didUpdateLocations locations:[AnyObject]!) {
        var location: CLLocation = locations[locations.count-1] as CLLocation
        
        if -location.timestamp.timeIntervalSinceNow > 2.0 {
            return
        }
        
        if (location as AnyObject? === nil) {
            if (firstUpdate) {
                firstUpdate = false
                firstLocation = location
            } else {
                secondLocation = location
                if (straight) {
                    distance = firstLocation.distanceFromLocation(secondLocation) * (metric ? 1 : 1.0936)
                } else {
                    distance += firstLocation.distanceFromLocation(secondLocation) * (metric ? 1 : 1.0936)
                    firstLocation = location
                }
            }
        }
        
        var units = metric ? "m" : "yds"
        estimateText.text = "\(distance) \(units)."
        oldDistance = distance
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        var destination: ResultsViewController = segue.destinationViewController as ResultsViewController
        destination.sendDistance(distance)
    }
}
